packages/web/src/store/application/ui/buy-audio/sagas.ts

```diff
@@ -0,0 +1,460 @@
+import {
+  Connection,
+  Keypair,
+  LAMPORTS_PER_SOL,
+  PublicKey,
+  Transaction
+} from '@solana/web3.js'
+import BN from 'bn.js'
+import JSBI from 'jsbi'
+import {
+  call,
+  select,
+  put,
+  takeEvery,
+  race,
+  delay
+} from 'typed-redux-saga/macro'
+
+import { BNWei } from 'common/models/Wallet'
+import { getAccountUser } from 'common/store/account/selectors'
+import { TOKEN_LISTING_MAP } from 'common/store/buy-audio/constants'
+import {
+  exchange,
+  exchangeAfterBalanceChange,
+  exchangeFailed,
+  exchangeSucceeded,
+  quote,
+  quoteFailed,
+  quoteSucceeded
+} from 'common/store/buy-audio/slice'
+import { increaseBalance } from 'common/store/wallet/slice'
+import { convertJSBIToUiString, weiToString } from 'common/utils/wallet'
+import AudiusBackend from 'services/AudiusBackend'
+import { waitForLibsInit } from 'services/audius-backend/eagerLoadUtils'
+
+const BALANCE_CHANGE_TIMEOUT_MS = 120_000 // 2 minutes
+const BALANCE_CHANGE_POLL_INTERVAL_MS = 5_000 // 5 second
+const WEI_DECIMALS = 18 // 18 decimals on ETH AUDIO
+const SPL_DECIMALS = 8 // 8 decimals on SPL AUDIO
+const SWAP_MIN_SOL = 0.05 // Minimum SOL balance needed to execute a Jupiter Swap
+const ERROR_CODE_SLIPPAGE = 6000 // Error code for when the swap fails due to specified slippage being exceeded
+let _jup: Jupiter
+
+const rpcList = [
+  'https://solana-api.projectserum.com',
+  'https://ssc-dao.genesysgo.net',
+  'https://mercuria-fronten-1cd8.mainnet.rpcpool.com'
+] // taken from jup.ag main site
+
+function convertWAudioToWei(amount: BN) {
+  const decimals = WEI_DECIMALS - SPL_DECIMALS
+  return amount.mul(new BN('1'.padEnd(decimals + 1, '0'))) as BNWei
+}
+
+/**
+ * Initializes Jupiter singleton if necessary and returns
+ * @returns {Jupiter} a Jupiter instance
+ */
+function* initJupiterIfNecessary() {
+  if (!_jup) {
+    let retries = 0
+    // Try all three RPC nodes if some fail to init for whatever reason
+    while (retries < rpcList.length) {
+      try {
+        _jup = yield* call(Jupiter.load, {
+          connection: new Connection(rpcList[retries], {
+            disableRetryOnRateLimit: true
+          }),
+          cluster: 'mainnet-beta',
+          restrictIntermediateTokens: true,
+          wrapUnwrapSOL: true,
+          routeCacheDuration: 5_000 // 5 seconds
+        })
+        if (_jup) {
+          break
+        }
+      } catch (e) {
+        console.warn('Jupiter failed to initialize with', rpcList[retries], e)
+        retries++
+      }
+    }
+  }
+  if (!_jup) {
+    throw new Error('Jupiter failed to initialize')
+  }
+  return _jup
+}
+/**
+ * Gets a quote from Jupiter for an exchange from inputTokenSymbol => outputTokenSymbol
+ * @returns the best quote including the RouteInfo
+ */
+function* doQuote({
+  inputTokenSymbol,
+  outputTokenSymbol,
+  inputAmount,
+  forceFetch,
+  slippage = 3,
+  padForSlippage = false
+}: ReturnType<typeof quote>['payload']) {
```

Comment:
This is fine to do, though ive found creating a type like type QuotePayload = {...}, type QuoteAction = PayloadAction<QuotePayload> and then can kinda share it around sagas and slices

Reply:
I think I like this more as then the type can remain encapsulated and not erroneously get suggested by autocomplete etc